### PR TITLE
feat: Improve e-sign request flow and email logic

### PR DESCRIPTION
**Pull request by Raju-kadel-27**
PR for issue #273

### CODE REVIEW

**Comment by Raju-kadel-27** [comment]
@dahal I think it's ready for review.

**Review by G3root — CHANGES_REQUESTED** [review]
Great work as always ❤️ 🚀 🚀 .

Here are some of my feedback and suggestions:

1. Save as a draft is not working.
2. `optional message` should be saved in the template model because ordered signing needs those values.
3. Maybe we can have an input for a custom `email subject` as well. We can save those optional values in the `template model` like `"optional message" => emailBody` and `emailSubject`. cc @dahal.
4. @dahal, should we allow `optional message` to be added to the draft as well? If so, rather than rendering a dropdown for `Save & Continue`, we can render a `Popover` that has all the inputs above and provide the user the option to complete or save as a draft.

File: src/trpc/routers/template-field-router/schema.ts [hunk 1]
```diff
@@ -23,6 +24,15 @@ export const ZodAddFieldMutationSchema = z.object({
       recipientId: z.string(),
     }),
   ),
+  emailPayload: z.object({
+    optionalMessage: z.string().optional(),
+    expiryDate: z.date().or(z.null()).or(z.string()).optional(),
+    documentName: z.string(),
```
[the remark this answers is not in the document]
Comment (Raju-kadel-27) [reply src/trpc/routers/template-field-router/schema.ts:30]: @G3root Sure ! Thanks for the feedback.

File: src/providers/template-field-provider.tsx [hunk 2]
```diff
@@ -41,8 +43,18 @@ const formSchema = z.object({
       }),
     )
     .nonempty(),
+
   recipient: z.string(),
   recipientColors: z.record(z.string()),
+  emailPayload: z.object({
+    optionalMessage: z.string().optional(),
+    expiryDate: z.date().or(z.null()).or(z.string()).optional(),
```
[the remark this answers is not in the document]
Comment (Raju-kadel-27) [reply src/providers/template-field-provider.tsx:51]: Currently we are just displaying it in email body only.
Do not have a clear idea.

File: package.json [hunk 3]
```diff
@@ -130,6 +130,6 @@
     "initVersion": "7.25.1"
   },
   "trigger.dev": {
-    "endpointId": "captable-A0-I"
+    "endpointId": "opencap-WhPZ"
```
Comment (dahal) [inline package.json:133]: Do we need to hardcode this? Each contributors and env will have different ids

File: public/avatar.svg
Comment (dahal) [inline public/avatar.svg:1]: We already have `/public/placeholders/user.svg`, please remove this.

File: public/signature.svg
Comment (dahal) [inline public/signature.svg:1]: Please remove this, and if we need re-usable icon, create one on `icons.tsx`

File: src/components/esign/optional-message-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    if (session) {
+      const email_payload = {
+        optionalMessage: optionalMessage ?? "",
+        completedOn: payload.completedOn,
+        documentName: payload.name,
+        company: {
+          name: payload.company.name,
+          logo:
+            payload.company.logo ??
+            "https://avatars.githubusercontent.com/u/163377635?s=48&v=4",
```
Comment (dahal) [inline src/components/esign/optional-message-modal.tsx:70]: Why is this hardcoded?

File: src/emails/EsignEmail.tsx [hunk 5]
```diff
@@ -1,67 +1,156 @@
...
+                    src={`https://jsx.email/assets/demo/vercel-arrow.png`}
+                    width="12"
+                    height="9"
+                    alt="invited you to"
+                  />
+                </Column>
+                <Column align="left">
+                  <Img
+                    className="rounded-full"
+                    src={"https://www.svgrepo.com/show/437226/signature.svg"}
```
Comment (dahal) [inline src/emails/EsignEmail.tsx:121]: Please dont hardcode these

File: src/emails/EsignEmail.tsx [hunk 6]
```diff
@@ -1,67 +1,156 @@
...
+            </Section>
+
+            <Section className="mx-auto mb-[32px] mt-[32px]">
               <Button
-                className="rounded bg-black px-5 py-3 text-center text-[12px] font-semibold text-white no-underline"
+                className="mx-[200px] rounded bg-black px-5 py-3 text-center text-[12px] font-semibold text-white no-underline"
                 href={signingLink}
               >
-                Sign document
+                Esign the template
```
Comment (dahal) [inline src/emails/EsignEmail.tsx:134]:
```suggestion
                Sign the document
```

File: src/emails/EsignEmail.tsx [hunk 7]
```diff
@@ -1,67 +1,156 @@
...
+                className="break-all text-blue-600 no-underline"
+              >
+                {signingLink}
+              </Link>
+            </Text>
 
-export default EsignEmail;
+            <Hr className="mx-0 my-[26px] w-full border border-solid border-neutral-200" />
+            <Text className="mx-auto text-center text-[12px] leading-[24px] text-[#666666]">
+              If you were not expecting this email, you can ignore this email.
```
Comment (dahal) [inline src/emails/EsignEmail.tsx:149]:
```suggestion
              Please ignore, if you were not expecting this email.
```

File: src/lib/utils.ts [hunk 8]
```diff
@@ -71,3 +71,27 @@ export function compareFormDataWithInitial<T extends Record<string, string>>(
...
+
+function getReadableMinute(minute: number) {
+  return minute < 10 ? `0${minute}` : `${minute}`;
+}
+
+function checkPrimeMeridian(hour: number) {
+  return hour > 12;
+}
+
+export function getSanitizedDateTime(datetime: string) {
```
Comment (dahal) [inline src/lib/utils.ts:87]: Please use `@/common/dayjs` to sanitize these dates and timestamps

File: src/providers/template-field-provider.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dahal) [reply src/providers/template-field-provider.tsx:51]: We will have the expiry on second iteration, lets remove for now

File: src/trpc/routers/template-field-router/procedures/add-fields.ts [hunk 9]
```diff
@@ -1,152 +1,108 @@
...
+        return {
+          success: false,
+          message: "E-signing has already completed among all parties.",
+        };
+      }
+      if (!user.email || !user.name || !companyLogo) {
+        return {
+          success: false,
+          message: "Email requires sender name , email and company logo.",
+        };
```
Comment (dahal) [inline src/trpc/routers/template-field-router/procedures/add-fields.ts:24]: Let's not make company logo a requirement to eSign. We have a placeholder we can use if company logo is not available.

`/public/placeholders/company.svg`

**Review by dahal — CHANGES_REQUESTED** [review]
Looking good, please go through the reviews and make necessary changes.

**Comment by dahal** [comment]
Here are my thoughts
> 2. `optional message` should be saved in the template model because ordered signing needs those values.

Yes, `optional message` should be on `template` model

> 3. Maybe we can have an input for a custom `email subject` as well. We can save those optional values in the `template model` like `"optional message" => emailBody` and `emailSubject`. cc @dahal.

For now, we can just limit to `message` and we can add subject, expiry and other optional fields on second iteration.

> 4. @dahal, should we allow `optional message` to be added to the draft as well? If so, rather than rendering a dropdown for `Save & Continue`, we can render a `Popover` that has all the inputs above and provide the user the option to complete or save as a draft.

We dont need to allow optional message on draft. Regarding popover, i have been experimenting it on Data room, i kind of like it. But for this use case, since modal is already implemented, lets limit message to `Send for signature` button.

File: package.json (hunk 3 above)
Comment (Raju-kadel-27) [reply package.json:133]: No, i will fix it.

**Comment by Raju-kadel-27** [comment]
@dahal @G3root  Fixed it.

**Pull request closed without merging**